client/components/data/query-post-counts/index.jsx

```diff
@@ -21,13 +21,15 @@ export default function QueryPostCounts( { siteId, type } ) {
 	const dispatch = useDispatch();
 
 	useEffect( () => {
-		dispatch( request( siteId, type ) );
+		if ( siteId && type ) {
+			dispatch( request( siteId, type ) );
+		}
 	}, [ dispatch, siteId, type ] );
 
 	return null;
 }
 
 QueryPostCounts.propTypes = {
-	siteId: PropTypes.number.isRequired,
-	type: PropTypes.string.isRequired,
+	siteId: PropTypes.number,
+	type: PropTypes.string,
```

Comment:
I think it was not a good idea to remove the `isRequired` here. It's still part of the component's contract that `siteId` and `type` should be specified. Rendering the component without them makes no sense and the consumer shouldn't do that. The type system (React prop types in our case) should warn about that.

The useful part of this PR is to avoid crashing on incorrect usage. But it's still incorrect usage and should be flagged as such.

Reply:
Yes, in hindsight I agree. Reinstated in #38288.

Reply:
Yes, in hindsight I agree. Reinstated in #38288.

Reply:
@jsnajdr  @sgomes It should no-op in that case, it's a real pain to have to have usage of `{ siteId && <QueryPostCounts /> }`. Add in other required fields and it ends up looking like `{ prop1 && prop2 && prop3 && <QueryComponent /> }`. Folks don't usually realize that state like currentSiteId are normally null on clean caches either.

Reply:
@gwwar When rendering a site-specific view, we usually use the `siteSelection` page.js handler to setup the `selectedSiteId` for us:
```js
page( '/foo/:site', siteSelection, renderFoo );
```
The handler will wait until all the sites are fetched, checks that the `:site` slug or ID is valid, will redirect to `/foo` (usually shows the site picker UI) if not, and gives control to `renderFoo` only after the `selectedSiteId` in Redux has the expected and valid value.

`renderFoo` can assume that valid `selectedSiteId` is a part of the contract it has with its caller and can rely on the contract being fulfilled.

> Folks don't usually realize that state like `currentSiteId` are normally `null` on clean caches either.

Therefore, this shouldn't be true for vast majority of views. `selectedSiteId` can be `null` for views that support an "All Sites" variant, and then the React UI component must be ready to handle that of course.

If there any views that experience a "flash of null selected site ID" on empty caches or in other circumstances, then it's likely a bug that should be fixed. I'd like to learn more about these.

`siteSelection` didn't always behave that well. Before #28449 was merged (a year ago), there indeed was a flash of `null` site ID, because the handler didn't wait before passing control -- instead, it caused `null` initial render and then Redux triggered a rerender when the sites got initialized from the REST endpoint. But these things are correct today.

Generally, functions and components should have reasonable contracts defined with their callers and the callers should be expected to fulfill these contracts. Not doing that can lead to a confused codebase where no one can ever trust that any prop or argument is valid and every function needs to have a lenghty prelude that validated every input. That's OK when handling untrusted input from outside world, but not for regular function calls in the program internals.

Reply:
Overall, I'm questioning the value of query components if the component needs to guard and validate against these cases anyway. It made sense in the past due to the number of lifecycle events, but now we can use a single `useEffect` function.

Reply:
I agree that hooks and `useEffect` greatly simplified the implemenation of the query components. But hooks can be used only in functional components, while wrapping them in a `Query` component makes them universal. The `Query` components can also be used in class components. We have thousands of these and will have them probably forever.